### PR TITLE
api: Ensure Timer is omitted when empty from Clock

### LINKED ISSUE #8844: The VirtualMachineInstance "foo" is invalid: spec.domain.clock.timer: Invalid value: "null": spec.domain.clock.timer in body must be of type object: "null"
**What happened**:

/area instancetype
/area api-server

[`Timer`](https://kubevirt.io/api-reference/master/definitions.html#_v1_timer) is listed as optional within [`Clock`](https://kubevirt.io/api-reference/master/definitions.html#_v1_clock) but isn't marked as `omitempty` in the actual schema:

https://github.com/kubevirt/kubevirt/blob/267dddd27d95d7e6a2ee6a87fdc5a42317e7f09e/staging/src/kubevirt.io/api/core/v1/schema.go#L857-L865

This can lead to validation failures if a `VirtualMachineInstanceSpec` is presented with a [`Clock`](https://kubevirt.io/api-reference/master/definitions.html#_v1_clock)  without a defined [`Timer`](https://kubevirt.io/api-reference/master/definitions.html#_v1_timer), for example:

```
spec:
  domain:
    clock:
      timezone: America/New_York
      utc: {}
```

**What you expected to happen**:
The `VirtualMachineInstance` should be admitted successfully.

**How to reproduce it (as minimally and precisely as possible)**:

This was first reported using the following `VirtualMachineClusterPreference` that has a `PreferredClockOffset`  but no `PreferredTimer`:

```
$ ./cluster-up/kubectl.sh kustomize https://github.com/kubevirt/common-instancetypes.git | ./cluster-up/kubectl.sh apply -f -
[..]
$ ./cluster-up/kubectl.sh apply -f - << EOF
---
apiVersion: instancetype.kubevirt.io/v1alpha2
kind: VirtualMachineClusterPreference
metadata:
  name: clocktest
spec:
  clock:
    preferredClockOffset:
      timezone: America/New_York
      utc: {}
---
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: clocktest
spec:
  instancetype:
    name: server.micro
  preference:
    name: clocktest
  running: true
  template:
    spec:
      domain:
        devices: {}
      volumes:      
      - containerDisk:
          image: registry:5000/kubevirt/cirros-container-disk-demo:devel
        name: containerdisk
EOF
$ ./cluster-up/kubectl.sh get vms/clocktest -o json | jq .status.conditions[1].message
selecting docker as container runtime
"Failure while starting VMI: VirtualMachineInstance.kubevirt.io \"clocktest\" is invalid: spec.domain.clock.timer: Invalid value: \"null\": spec.domain.clock.timer in body must be of type object: \"null\""
```

The above is equal to the following `VirtualMachineInstance` that also reproduces the validation error:
```
$ ./cluster-up/kubectl.sh apply -f - << EOF
---
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  name: clocktestvmi
spec:
  domain:
    clock:
      timezone: America/New_York
      utc: {}
    cpu:
      cores: 1
      model: host-model
      sockets: 1
      threads: 1
    devices:
      disks:
      - disk:
          bus: sata
        name: containerdisk
      interfaces:
      - bridge: {}
        name: default
    features:
      acpi:
        enabled: true
    machine:
      type: q35
    memory:
      guest: 256Mi
    resources:
      requests:
        memory: 256Mi
  networks:
  - name: default
    pod: {}
  volumes:
  - containerDisk:
      image: registry:5000/kubevirt/cirros-container-disk-demo:devel
      imagePullPolicy: IfNotPresent
    name: containerdisk
EOF
The VirtualMachineInstance "clocktestvmi" is invalid: spec.domain.clock.timer: Invalid value: "null": spec.domain.clock.timer in body must be of type object: "null"
```

**Additional context**:

[Bug 2145126](https://bugzilla.redhat.com/show_bug.cgi?id=2145126) - Cant start VM with "clock" virtualMachinePreference
https://bugzilla.redhat.com/show_bug.cgi?id=2145126

**Environment**:
- KubeVirt version (use `virtctl version`): N/A
- Kubernetes version (use `kubectl version`): N/A
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by lyarwood**
/area instancetype
/area api-server

**What this PR does / why we need it**:

As set out in bug #8844 when `Timer` was previously empty it was not omitted from `Clock`  causing validation failures to be raised as an object was expected. This PR now ensures that `Timer` is omitted from `Clock` when empty, avoiding the validation failure.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8844

**Special notes for your reviewer**:

**Release note**:

```release-note
An empty `Timer` is now correctly omitted from `Clock` fixing bug #8844.
```

### CODE REVIEW

**Review by vasiliy-ul — APPROVED** [review]

**Comment by lyarwood** [comment]
> Don't we need a release note for this ? You're updating the API. @lyarwood

I can add one, wasn't sure as this is technically just a bug fix.

**Review by maiqueb — APPROVED** [review]

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go [hunk 1]
```diff
@@ -138,6 +138,29 @@ var _ = Describe("Validating VMICreate Admitter", func() {
...
+		ar := &admissionv1.AdmissionReview{
+			Request: &admissionv1.AdmissionRequest{
+				Resource: webhooks.VirtualMachineInstanceGroupVersionResource,
+				Object: runtime.RawExtension{
+					Raw: vmiBytes,
+				},
+			},
+		}
+		resp := vmiCreateAdmitter.Admit(ar)
+		Expect(resp.Allowed).To(BeTrue())
```
Comment (maiqueb) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go:161]: nit: this could be inlined.

(feel free to ignore).

**Comment by xpivarc** [comment]
/cc @rthallisey 
FYI

**Comment by rthallisey** [comment]
/cc @alaypatel07
Looks ok from API backward compatibility perspective.

**Comment by alaypatel07** [comment]
+1, lgtm upgrade/downgrade will both continue to work with this changes.

**Review by xpivarc — COMMENTED** [review]
Good catch!
/approve

**Pull request merged**